File: 2.Strings/C/removeDuplicates.c
/**
 * Remove Duplicates from String
 * 
 * This program removes duplicate characters from a string, preserving the
 * order of the first occurrence of each character.
 * 
 * Algorithm:
 * 1. Handle edge cases (NULL string or string with length < 2)
 * 2. Keep track of unique characters seen so far using a 'tail' index
 * 3. For each character in the string, check if it already exists in the processed part
 * 4. If it's a new character, add it to the result at the 'tail' position
 * 5. Terminate the string at the new length
 * 
 * Time Complexity: O(n²) where n is the length of the string
 * Space Complexity: O(1) as we modify the string in-place
 * 
 * Input: A string - "programming"
 * Output: String with duplicates removed - "progamin"
 */

#include <stdio.h>
#include <string.h>

/**
 * removeDuplicates - Remove duplicate characters from a string in-place
 * @str: The input string to modify
 * 
 * This function removes all duplicate characters from the string while
 * preserving the order of the first occurrence of each character.
 */
void removeDuplicates(char *str) {
    // Handle NULL string
    if (str == NULL)
        return;
    
    // Get string length
    int len = strlen(str);
    
    // No duplicates possible in strings of length 0 or 1
    if (len < 2)
        return;
    
    // 'tail' keeps track of the position where the next unique character will be placed
    // We assume the first character is always unique, so start with tail = 1
    int tail = 1;
    
    // Process each character starting from the second one
    for (int i = 1; i < len; i++) {
        int j;
        // Check if the current character already exists in the processed part
        for (j = 0; j < tail; j++) {
            if (str[i] == str[j])
                break;  // Duplicate found
        }
        
        // If we went through the entire processed part without finding a duplicate
        if (j == tail) {
            str[tail] = str[i];  // Add this unique character to the result
            tail++;  // Move the tail pointer forward
        }
    }
    
    // Terminate the string at the new length
    str[tail] = '\0';
}

int main() {
    // Test string
    char str[] = "programming";
    
    // Display the original string
    printf("Original string: %s\n", str);
    
    // Remove duplicates and display the result
    removeDuplicates(str);
    printf("After removing duplicates: %s\n", str);
    
    return 0;
}

File: 2.Strings/java/RemoveDuplicates.java
import java.util.*;

/**
 * RemoveDuplicates
 * 
 * This class provides methods to remove duplicate characters from a string while
 * preserving the order of the first occurrence of each character.
 * 
 * Three implementations are provided:
 * 1. Using LinkedHashSet (removeDuplicates)
 * 2. Manual implementation using StringBuilder and indexOf (removeDuplicatesManual)
 * 3. Using Java 8 Streams (removeDuplicatesStream)
 * 
 * Time Complexity:
 * - LinkedHashSet: O(n) where n is the length of the string
 * - Manual: O(n²) due to indexOf operation for each character
 * - Streams: O(n) for the distinct operation
 * 
 * Space Complexity: O(n) for all methods as they create new strings
 * 
 * Input: A string with potential duplicate characters
 * Output: A string with duplicates removed, preserving the order of first occurrence
 */
public class RemoveDuplicates {
    /**
     * Removes duplicate characters using LinkedHashSet.
     * 
     * Algorithm:
     * 1. Create a LinkedHashSet to maintain insertion order and uniqueness
     * 2. Add each character from the string to the set
     * 3. Build a new string from the set's elements
     * 
     * @param str The input string
     * @return A string with duplicates removed
     */
    public static String removeDuplicates(String str) {
        // LinkedHashSet maintains insertion order while ensuring uniqueness
        LinkedHashSet<Character> set = new LinkedHashSet<>();
        
        // Add each character to the set (duplicates are automatically ignored)
        for (char c : str.toCharArray()) {
            set.add(c);
        }
        
        // Build the result string from the set
        StringBuilder sb = new StringBuilder();
        for (Character c : set) {
            sb.append(c);
        }
        return sb.toString();
    }
    
    /**
     * Removes duplicate characters manually using StringBuilder and indexOf.
     * 
     * Algorithm:
     * 1. Handle edge cases (null or single character strings)
     * 2. Create a StringBuilder to store the result
     * 3. For each character, check if it already exists in the result
     * 4. If not found, append it to the result
     * 
     * @param str The input string
     * @return A string with duplicates removed
     */
    public static String removeDuplicatesManual(String str) {
        // Handle edge cases
        if (str == null || str.length() <= 1) {
            return str;
        }
        
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < str.length(); i++) {
            char current = str.charAt(i);
            // indexOf returns -1 if the character is not found
            if (result.indexOf(String.valueOf(current)) == -1) {
                result.append(current);
            }
        }
        
        return result.toString();
    }
    
    /**
     * Removes duplicate characters using Java 8 Streams.
     * 
     * Algorithm:
     * 1. Convert the string to an IntStream of character codes
     * 2. Use distinct() to remove duplicates
     * 3. Map each integer back to its character representation
     * 4. Concatenate the characters to form the result string
     * 
     * @param str The input string
     * @return A string with duplicates removed
     */
    public static String removeDuplicatesStream(String str) {
        return str.chars()                      // Convert to IntStream
                 .distinct()                    // Remove duplicates
                 .mapToObj(ch -> String.valueOf((char)ch))  // Convert back to String
                 .reduce("", String::concat);  // Concatenate all characters
    }
    
    /**
     * Main method to test all three implementations with various examples.
     * 
     * @param args Command line arguments (not used)
     */
    public static void main(String[] args) {
        // Test cases with different patterns
        String[] testStrings = {
            "programming",    // Multiple duplicates
            "hello world",   // Duplicates with spaces
            "java streams",  // Multiple words
            "aabbccdd"       // Consecutive duplicates
        };
        
        // Test each string with all three methods
        for (String str : testStrings) {
            System.out.println("\nOriginal string: " + str);
            System.out.println("After removing duplicates (Set): " + 
                             removeDuplicates(str));
            System.out.println("After removing duplicates (Manual): " + 
                             removeDuplicatesManual(str));
            System.out.println("After removing duplicates (Stream): " + 
                             removeDuplicatesStream(str));
        }
    }
}